### PR TITLE
feature: Restrict Kubechecks to a Single Namespace for App Watcher  

### PR DESCRIPTION
**Pull request by Greyeye**
# Changes

1.  **Namespace Restriction for Kubechecks**:
    
    *   **Current Behavior**: Kubechecks monitors all ArgoCD applications and ApplicationSet CRDs across the entire cluster. This is not ideal when multiple ArgoCD instances are running in a single cluster (e.g., for testing purposes).
    *   **New Behavior**: Introduces the ability to specify a namespace for ArgoCD. Kubechecks will now watch ArgoCD CRDs only within the specified namespace.
        *   Added a new flag `argocd-api-namespace` and a corresponding server config `ArgoCDNameSpace`.
        *   Modified `NewApplicationInformer` to target the specified `argocdnamespace`.
2.  **Improvement to `dumpcrd`**:
    
    *   `dumpcrd` can now run without requiring a local `kubectl` binary.
3.  **Modification to GitHub Webhook Check**:
    
    *   The webhook check for GitHub has been updated to validate against the payload URL instead of the hook's API URL.
4.  **OpenAI ChatGPT Model Update**:
    
    *   Updated to use the `GPT-4o` model.
5.  GitHub lib updated to v62
    
# Localdev Changes  

1.  **Earthly File Renaming**:
    
    *   Renamed the local earthly file to `earthly.sh` to minimize confusion.
2.  **TiltFile Fix**:
    
    *   Updated `TiltFile` to read the `.tools-version` file directly, preventing failures during the earthly build.
    *   Updated to read `.secret` instead of `.env`; as secrets env vars are read from the .secret
3.  **Values.yaml Update**:
    
    *   Renamed `KUBECHECKS_ARGOCD_NAMESPACE` to `KUBECHECKS_ARGOCD_API_NAMESPACE` in `localdev/kubechecks/values.yaml`.
4.  **Configuration Change**:
    
    *   Set `KUBECHECKS_ENABLE_CONFTEST` to `false` as the `./policy` path is not valid.

### CODE REVIEW

File: .github/workflows/on_pull_request.yaml [hunk 1]
```diff
@@ -15,7 +15,7 @@ on:
       - synchronize
 
 env:
-  FS_TAG: 0.0.0-pr${{ github.event.pull_request.number }}
+  FS_TAG: 0.0.0-pr${{ github.event.pull_request.number }}-${{ github.sha }}
```
Comment (djeebus) [inline .github/workflows/on_pull_request.yaml:18]: why add the sha here? The only problem is that the [delete tag script](https://github.com/zapier/kubechecks/blob/green-fix/.github/workflows/on_pull-request_closed.yaml#L30) stops working, and you can't use a wildcard there. could we fix the reason for this w/ a `imagePullPolicy: Always` somewhere?

File: Tiltfile [hunk 2]
```diff
@@ -138,12 +138,59 @@ test_go(
...
+        line = lines[linenumber]
+        parts = line.split("#", 1)
+        if len(parts) == 2:
+            line = parts[0]
+        line = line.strip()
+        if line == "":
+            continue
+        parts = line.split(' ', 1)
+        tools[parts[0].strip()] = parts[1].strip()
+    tools['githead'] = str(get_git_head())
```
Comment (djeebus) [inline Tiltfile:169]: why shove githead into tools? you could do `githead = get_git_head()` at the global level, then reference variable

File: cmd/container.go [hunk 3]
```diff
@@ -60,6 +60,8 @@ func newContainer(ctx context.Context, cfg config.ServerConfig, watchApps bool)
 
 			go ctr.ApplicationWatcher.Run(ctx, 1)
 		}
+	} else {
+		log.Debug().Msgf("not monitoring applications, MonitorAllApplications: %+v", cfg.MonitorAllApplications)
```
Comment (djeebus) [inline cmd/container.go:64]: using `Info` here might be useful, to make it clear to users what is happening and (maybe) why their applications aren't being monitored the way they expect

File: cmd/root.go [hunk 4]
```diff
@@ -56,8 +56,13 @@ func init() {
 			withDefault("gitlab"))
 	stringFlag(flags, "vcs-token", "VCS API token.")
 	stringFlag(flags, "argocd-api-token", "ArgoCD API token.")
-	stringFlag(flags, "argocd-api-server-addr", "ArgoCD API Server Address.", newStringOpts().withDefault("argocd-server"))
+	stringFlag(flags, "argocd-api-server-addr", "ArgoCD API Server Address.",
+		newStringOpts().
+			withDefault("argocd-server"))
 	boolFlag(flags, "argocd-api-insecure", "Enable to use insecure connections to the ArgoCD API server.")
+	stringFlag(flags, "argocd-api-namespace", "ArgoCD namespace where the application watcher will read Custom Resource Definitions (CRD).",
```
Comment (djeebus) [inline cmd/root.go:63]: I think, technically, the CRD is the [schema definition](https://github.com/argoproj/argo-cd/blob/master/manifests/crds/application-crd.yaml), I think we want to say something like "Application and ApplicationSet resources"

File: pkg/aisummary/diff_summary.go [hunk 5]
```diff
@@ -17,7 +17,7 @@ func (c *OpenAiClient) SummarizeDiff(ctx context.Context, appName, diff string)
 	ctx, span := tracer.Start(ctx, "SummarizeDiff")
 	defer span.End()
 
-	model := openai.GPT4Turbo0125
+	model := openai.GPT4o
```
Comment (djeebus) [inline pkg/aisummary/diff_summary.go:20]: :open_mouth:

File: pkg/app_watcher/app_watcher_test.go [hunk 6]
```diff
@@ -8,12 +8,18 @@ import (
...
 	"github.com/zapier/kubechecks/pkg/appdir"
 )
 
-func initTestObjects() *ApplicationWatcher {
+func initTestObjects(t *testing.T) *ApplicationWatcher {
+	cfg, err := config.New()
+	if err != nil {
+		// Handle the error appropriately, e.g., log it or fail the test
+		t.Fatalf("Failed to initialize config: %v", err)
+	}
```
Comment (djeebus) [inline pkg/app_watcher/app_watcher_test.go:22]: `require.NoError(err)` would be a little more succinct.

**Review by djeebus — APPROVED** [review]

File: .github/workflows/on_pull_request.yaml (hunk 1 above)
Comment (Greyeye) [reply .github/workflows/on_pull_request.yaml:18]: @djeebus, when I tested it, the pull policy `always` didn't work. When you make numerous changes to the PR, our cache pulls the old one, not the latest. I had to force this for EKS to pull the different tag. I'll revert it for now, but not sure there is any better solution.

File: Tiltfile (hunk 2 above)
Comment (Greyeye) [reply Tiltfile]: changed it, it is now calling get_git_head() out side.

File: cmd/container.go (hunk 3 above)
Comment (Greyeye) [reply cmd/container.go]: changed to info

File: pkg/app_watcher/app_watcher_test.go (hunk 6 above)
Comment (Greyeye) [reply pkg/app_watcher/app_watcher_test.go]: changed to 
assert.NoError(t, err, "failed to create config")

File: .github/workflows/on_pull_request.yaml (hunk 1 above)
Comment (Greyeye) [reply .github/workflows/on_pull_request.yaml]: reverted it

**Review by djeebus — APPROVED** [review]

**Review by polyrain — APPROVED** [review]
Approved, thanks!

**Pull request merged**